### PR TITLE
Tolerate empty GroupVersions when syncing

### LINKED ISSUE #1951: Flux chokes on empty Custom Metrics GroupVersion
Running into a strange problem. 

I have a release that is managed by Flux+Helm. So far so good. Recently pushed an updated docker image and now Flux can't deploy it. 

What's really concerning is, Flux pushes a commit to the Gitops repo saying the new release was rolled out, even though the operation clearly fails. 

Running `fluxctl sync`:
```sh
fluxctl sync
Synchronizing with <redacted>
HEAD of master is ed3ba29
Waiting for ed3ba29 to be applied ...
Error: timeout
```

Some of the errors I see are `collating resources in cluster for sync: not found` and `couldn't get resource list for custom.metrics.k8s.io/v1beta1: <nil>`. Not super helpful. 

Log:
```
2019-04-17T16:18:07.202427131Z ts=2019-04-17T16:18:07.202281634Z caller=loop.go:111 component=sync-loop jobID=fb24233b-779c-3020-99db-647abc155bac state=in-progress
2019-04-17T16:18:09.15816782Z ts=2019-04-17T16:18:09.158030323Z caller=loop.go:123 component=sync-loop jobID=fb24233b-779c-3020-99db-647abc155bac state=done success=true
2019-04-17T16:18:11.084675671Z ts=2019-04-17T16:18:11.084497151Z caller=loop.go:103 component=sync-loop event=refreshed url=<redacted> branch=master HEAD=ed3ba298cdc026dbb635eaf731548395271bb233
2019-04-17T16:18:31.21213776Z ts=2019-04-17T16:18:31.211961076Z caller=images.go:18 component=sync-loop msg="polling images"
2019-04-17T16:18:31.335243305Z ts=2019-04-17T16:18:31.335087864Z caller=main.go:199 type="internal kubernetes error" err="couldn't get resource list for custom.metrics.k8s.io/v1beta1: <nil>"
2019-04-17T16:18:31.385604796Z ts=2019-04-17T16:18:31.385439156Z caller=images.go:109 component=sync-loop workload=staging:helmrelease/decompress container=chart-image repo=<redacted> pattern=glob:staging* current=<redacted>:staging-v1.3.1-2 info="added update to automation run" new=<redacted>:staging-v1.3.1-3 reason="latest staging-v1.3.1-3 (2019-04-16 16:47:58.788602027 +0000 UTC) > current staging-v1.3.1-2 (2019-04-12 23:52:13.588390498 +0000 UTC)"
2019-04-17T16:18:31.385627972Z ts=2019-04-17T16:18:31.385537557Z caller=loop.go:111 component=sync-loop jobID=cf4c96e4-54c6-0777-b868-9ef41c8180ee state=in-progress
2019-04-17T16:18:31.484304234Z ts=2019-04-17T16:18:31.484182549Z caller=releaser.go:58 component=sync-loop jobID=cf4c96e4-54c6-0777-b868-9ef41c8180ee type=release updates=0
2019-04-17T16:18:31.484331638Z ts=2019-04-17T16:18:31.484218832Z caller=releaser.go:60 component=sync-loop jobID=cf4c96e4-54c6-0777-b868-9ef41c8180ee type=release exit="no images to update for services given"
2019-04-17T16:18:31.523865111Z ts=2019-04-17T16:18:31.523740767Z caller=loop.go:121 component=sync-loop jobID=cf4c96e4-54c6-0777-b868-9ef41c8180ee state=done success=false err="no changes made in repo"
2019-04-17T16:18:33.065109033Z ts=2019-04-17T16:18:33.064978047Z caller=loop.go:103 component=sync-loop event=refreshed url=<redacted> branch=master HEAD=ed3ba298cdc026dbb635eaf731548395271bb233
2019-04-17T16:18:37.929537537Z ts=2019-04-17T16:18:37.929414905Z caller=loop.go:210 component=sync-loop err="collating resources in cluster for sync: not found"
2019-04-17T16:18:37.969697408Z ts=2019-04-17T16:18:37.969557523Z caller=loop.go:90 component=sync-loop err="collating resources in cluster for sync: not found"
```

The image I want rolled is staging-v1.3.1-3, the currently running one is staging-v1.3.1-2. 

The upgrade from v1.3.1-1 to v1.3.1-2 went off without a hitch, which was with an older releae of Flux. 

I am now running the fresh-est release of Flux: `docker.io/weaveworks/flux:1.12.0` and `docker.io/weaveworks/helm-operator:0.8.0`

Any ideas?

### PR DESCRIPTION
**Pull request by 2opremio**
Also:
* Improve error reporting
* Use ServerGroupsAndResources instead of ServerGroups (deprecated)

Fixes #1951

### CODE REVIEW

File: cluster/kubernetes/sync.go [hunk 1]
```diff
@@ -198,9 +199,20 @@ func (c *Cluster) getAllowedResourcesBySelector(selector string) (map[string]*ku
...
+			return nil, err
+		}
+		for gv, e := range discErr.Groups {
+			// Tolerate empty GroupVersions due to e.g. misconfigured custom metrics
+			if e.Error() != fmt.Sprintf("Got empty response for: %v", gv) {
+				return nil, err
+			}
+		}
+		c.logger.Log("warn", err,
+			"impact", "ignoring error, please check your cluster configuration, GroupVersions should not be empty")
```
Comment (2opremio) [inline cluster/kubernetes/sync.go]: I am on the fence about this. It feels like telling users off without any real advantage.

I think it's probably fine to be silent about it.

Comment (squaremo) [reply cluster/kubernetes/sync.go]: Maybe just log it once? Or if it's rare, and there's some characteristic symptom of it failing, we can be silent here and add a troubleshooting doc entry instead.

File: daemon/loop.go [hunk 2]
```diff
@@ -207,9 +207,9 @@ func (d *Daemon) doSync(logger log.Logger, lastKnownSyncTagRev *string, warnedAb
 
 	var resourceErrors []event.ResourceError
 	if err := fluxsync.Sync(syncSetName, allResources, d.Cluster); err != nil {
-		logger.Log("err", err)
 		switch syncerr := err.(type) {
 		case cluster.SyncError:
+			logger.Log("err", err)
```
Comment (squaremo) [inline daemon/loop.go]: Not quite sure of the motivation for this change -- what errors were being logged that shouldn't be?

Comment (2opremio) [reply daemon/loop.go]: The problem was that the error was logged twice, by this function and the caller

Comment (squaremo) [reply daemon/loop.go]: Got it, ta

File: cluster/kubernetes/sync.go (hunk 1 above)
Comment (hiddeco) [reply cluster/kubernetes/sync.go]: As we talked about reducing the log clutter a while ago and the reports about this problem thus far have been Datadog metric users (so _I think_ the problem is very niche), I agree with you @fons and it is probably fine to be silent about it.

Comment (2opremio) [reply cluster/kubernetes/sync.go]: I will silence it then

**Pull request merged**